### ISSUE TITLE
Can't restart kernel

### ISSUE THREAD

**Issue by MasonProtter** [dzop/emacs-jupyter#28]
When I try to do `M-x jupyter-repl-restart-kernel` I get the following error:
```
Wrong type argument: jupyter-kernel-client, nil, client
```

However, `C-c C-r` does succeed in restarting the kernel.

**Comment by dzop** [comment]
Maybe it has to do with the buffer that is current whenever you do `M-x jupyter-repl-restart-kernel` vs `C-c C-r` since the error looks to be coming from the `jupyter-set` function which ensures that its `client` argument is valid.

Can you verify that `jupyter-current-client` is non-nil before the 

```emacs-lisp
(jupyter-set jupyter-current-client 'jupyter-include-other-output t)
```

line in `jupyter-repl-restart-kernel` and show the value of `current-buffer` for both cases, e.g. by  putting

```emacs-lisp
(message "buffer %s client %s" (current-buffer) jupyter-current-client)
```

before the line mentioned above.

**Comment by MasonProtter** [comment]
>Can you verify that `jupyter-current-client` is non-nil before the

```emacs-lisp
(jupyter-set jupyter-current-client 'jupyter-include-other-output t)
```

> line in `jupyter-repl-restart-kernel` and show the value of `current-buffer` for both cases, e.g. by putting

Sorry, I'm not sure what you mean here. Do you want me to look at the source code for the package, or was I supposed to execute `(jupyter-set jupyter-current-client 'jupyter-include-other-output t)` somewhere in my startup? 


_____

Also, I just noticed that when I do `C-c C-r`, it displays a message saying that its restarting the kernel but actually, the session appears to keep going uninterrupted. I still have all my variable definitions and whatnot from the previous session.

**Comment by dzop** [comment]
Yes, I meant to go into the source code and modify `jupyter-repl-restart-kernel` by putting the `message` line above the `jupyter-set` line, call `eval-defun`, try the two cases, and then show the value of what was printed if that is OK.

It does look like restarting the kernel has some issues after trying it myself. The error I'm experiencing is not the exact same, but may be related. Can you also show the backtrace you get when the error occurs by calling `toggle-debug-on-error` before attempting to restart?

**Comment by MasonProtter** [comment]
Okay, done. When I call `M-x jupyter-repl-restart-kernel` the message output is
```
buffer BCS_PH.org client nil
helm-M-x: Wrong type argument: jupyter-kernel-client, nil, client
```
and the debugger output is
```
Debugger entered--Lisp error: (wrong-type-argument jupyter-kernel-client nil client)
  signal(wrong-type-argument (jupyter-kernel-client nil client))
  jupyter-set(nil jupyter-include-other-output t)
  jupyter-repl-restart-kernel(nil)
  funcall-interactively(jupyter-repl-restart-kernel nil)
  call-interactively(jupyter-repl-restart-kernel record nil)
  command-execute(jupyter-repl-restart-kernel record)
  helm-M-x(nil #("jupyter-repl-restart-kernel" 0 27 (match-part "jupyter-repl-restart-kernel")))
  funcall-interactively(helm-M-x nil #("jupyter-repl-restart-kernel" 0 27 (match-part "jupyter-repl-restart-kernel")))
  call-interactively(helm-M-x nil nil)
  command-execute(helm-M-x)
```

When I do `C-c C-r` I get a gigantic message output which I put here: https://pastebin.com/JxzMXRfs and no error.

**Comment by dzop** [comment]
Ah OK, yeah `M-x jupyter-repl-restart-kernel` won't work in an `org-mode` buffer. `C-c C-r` works because it was bound to the `jupyter-repl-restart-kernel` command using `jupyter-org-define-key` which augments the command to look at the surrounding src-block to get the right client. Using `M-x jupyter-repl-restart-kernel` just calls the command directly and there is no client that is associated with the entire `org-mode` buffer, i.e. `jupyter-current-client` is nil in the `org-mode` buffer.

Also about the `session appears to keep going uninterrupted` issue. I think I found the problem. The Julia kernel doesn't seem to respond to a shutdown request (at least for me on Julia 1.0) and, in this case, the kernel wasn't being forcibly shutdown. I have fixed this on the latest master.

**Comment by MasonProtter** [comment]
So how can I kill an active session when in org mode? Do I have to open a repl and then kill that, or what?

**Comment by dzop** [comment]
If `point` is on a src-block connected to the session, doesn't `C-c C-r` work?

**Comment by MasonProtter** [comment]
No.

**Comment by dzop** [comment]
To also have `M-x jupyter-repl-restart-kernel` work what we can do is have  the command prompt for a REPL buffer to restart if the current buffer doesn't have a client associated with it. This way it will also work in `org-mode`, you will just have to select the right REPL buffer.

**Comment by dzop** [comment]
Closing since I believe issues with restarting the REPL have been fixed in 1f5c7743d534929efaad7592762f241700bc1f6c.
